### PR TITLE
Fix issue #7232: Do not mask exceptions in objective func evaluation

### PR DESCRIPTION
**Pull request by jllanfranchi**
* evaluate `func(x)` outside try/except block
* catch a particular exception (`TypeError`), rather than _all_ exceptions in check

### CODE REVIEW

File: scipy/optimize/slsqp.py [hunk 1]
```diff
@@ -364,9 +364,10 @@ def cjac(x, *args):
         if mode == 0 or mode == 1:  # objective and constraint evaluation requird
 
             # Compute objective function
+            fx = func(x)
             try:
-                fx = float(np.asarray(func(x)))
-            except:
+                fx = float(np.asarray(fx))
+            except TypeError:
```
Comment (rgommers) [inline scipy/optimize/slsqp.py]: ``float('somestring')`` gives ``ValueError``, so that should be caught as well.

**Comment by rgommers** [comment]
Looks good to me once you catch ``ValueError`` as well.

**Comment by jllanfranchi** [comment]
Many thanks; updated accordingly.

**Comment by rgommers** [comment]
Merged, thanks @jllanfranchi

**Pull request merged**